[AMDGPU] Split vgpr regalloc pipeline

cdevadas opened the pull request:
Allocating wwm-registers and per-thread VGPR operands
together imposes many challenges in the way the
registers are reused during allocation. There are
times when regalloc reuses the registers of regular
VGPRs operations for wwm-operations in a small range
leading to unwantedly clobbering their inactive lanes
causing correctness issues that are hard to trace.

This patch splits the VGPR allocation pipeline further
to allocate wwm-registers first and the regular VGPR
operands in a separate pipeline. The splitting would
ensure that the physical registers used for wwm
allocations won't take part in the next allocation
pipeline to avoid any such clobbering.

cdevadas committed:
[AMDGPU] Split VGPR regalloc pipeline

Allocating wwm-registers and regular VGPR operands
together imposes many challenges in the way the
registers are reused during allocation. There are
times when regalloc reuses the registers of regular
VGPRs operations for wwm-operations in a small range
leading to unwantedly clobbering their inactive lanes
causing correctness issues which are hard to trace.

This patch splits the VGPR allocation pipeline further
to allocate wwm-registers first and the regular VGPR
operands in a separate pipeline. The splitting would
ensure that the physical registers used for wwm
allocations won't taken part in the next allocation
pipeline to avoid any such clobbering.

cdevadas commented:
This patch would replace the PR https://github.com/llvm/llvm-project/pull/86012. The two generic features introduced earlier were strongly discouraged. This patch splits the VGPR allocation further to handle the wwm-allocation in a target-specific way.

cdevadas commented:
Some tests I marked XFAIL for now. They should be fixed.

cmc-rep reviewed (commented):
Looks good to me

arsenm commented on llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:
```diff
@@ -32,12 +33,18 @@ using MBBVector = SmallVector<MachineBasicBlock *, 4>;
+static cl::opt<unsigned> MaxNumVGPRsForWwmAllocation(
+    "amdgpu-num-vgprs-for-wwm-alloc",
+    cl::desc("Max num VGPRs for whole-wave register allocation."),
+    cl::ReallyHidden, cl::init(5));
```
This seems really small and odd for a maximum?

cdevadas replied:
I see that. I initially wanted this value between 5 - 7 based on the experience with SGPR spilling on heavy workloads earlier. In the internal testing, I used 5 and 10 for the upper cap and they both worked without any issues. I should have kept the higher value here. Will change it to 10.

arsenm commented on llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:
```diff
@@ -259,51 +271,86 @@ bool SILowerSGPRSpills::spillCalleeSavedRegs(
     }
   }
+
+  assert(I == NumRegs &&
+         "Failed to find enough VGPRs for whole-wave register allocation");
```
This probably needs to be report_fatal_error (or better, emitError and just pick a register). You can easily run into this with some asm

cdevadas replied:
I prefer report_fatal_error. Picking any register wouldn't help here. It will anyway be a bad codegen.

arsenm replied:
Yes, it is an invalid compile but it allows you to properly return to the frontend with an error pointing at the source location

perlfu commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1387,11 +1455,20 @@ bool GCNPassConfig::addRegAssignAndRewriteFast() {
+
+  // To Allocate wwm registers used in whole quad mode operations (for pixel
+  // shaders).
   addPass(&SIPreAllocateWWMRegsID);
```
I think the comment in the code is misleading since WWM registers pre-allocated are used in all shader types (not just PS) for atomic optimization and subgroup intercommunication.

cdevadas replied:
I will mention the shaders instead of PS alone.

cdevadas replied:
> My comment here maybe should be a follow up to this patch: SIPreAllocateWWMRegs perhaps should be renamed to SIPreAllocateWQMRegsInPS, and its functionality should be limited to WQM registers. 

I would rather not touch it in this patch. These passes will anyway get a revamp in your WIP allocation patches. Let's fix the pass name at that point.

arsenm commented on llvm/lib/Target/AMDGPU/SILowerSGPRSpills.cpp:
```diff
@@ -349,8 +349,12 @@ void SILowerSGPRSpills::determineRegsForWWMAllocation(MachineFunction &MF,
+  if (I != NumRegs) {
+    // Reserve an arbitrary register and report the error.
+    TRI->markSuperRegs(RegMask, AMDGPU::VGPR0);
+    MF.getFunction().getContext().emitError(
+        "can't find enough VGPRs for wwm-regalloc");
```
Should add a test for this (improving all of these RA-related error handling is on my todo list)

cdevadas replied:
Yes, writing one and also, fixing the tests I made XFAIL.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUReserveWWMRegs.cpp:
```diff
@@ -0,0 +1,103 @@
+    }
+  }
+
+  // Reset the renamable flag for MOs involving wwm-regs to get rid of the MIR
+  // Verifier error.
```
What verifier error? State what is invalid, not that there is a verifier error

cdevadas replied:
After wwm-regalloc is done, we reserve those VGPRs before vgpr-regalloc. The reserved registers shouldn't have the renamable flag. I will rephrase the comment.

cdevadas commented:
Added a negative test llvm/test/CodeGen/AMDGPU/wwm-regalloc-error.ll to capture the error during wwm-regalloc.
Fixed all tests earlier marked XFAIL.
Taken care of the review comments.

arsenm commented on llvm/test/CodeGen/AMDGPU/wwm-regalloc-error.ll:
```diff
@@ -0,0 +1,17 @@
+  ret void
+}
+
+attributes #0 = { nounwind "amdgpu-num-vgpr"="2"}
```
Drop the attribute, we should remove it eventually and it should be redundant with -stress-regalloc

cdevadas replied:
The `stress-regalloc` is a niche way to restrict only the AllocationOrder. It is visible only to RegisterClassInfo.cpp design and not useful to constrain the register numbers elsewhere. So, `getMaxNumVGPRs` doesn't incorporate it at all. I should otherwise clobber all available VGPRs via inline ASM to make them unavailable.

cdevadas replied:
I've been thinking of moving it to the stress-regalloc to MRI so that targets can access them using a helper function throughout their codegen. But that is altogether a different patch.

arsenm commented on llvm/test/CodeGen/AMDGPU/wwm-regalloc-error.ll:
```diff
@@ -0,0 +1,17 @@
+; RUN: not llc -mtriple=amdgcn-amd-amdhsa -mcpu=gfx900 -mattr=-flat-for-global -stress-regalloc=3 -filetype=null %s 2>&1 | FileCheck %s
```
Remove -mattr=-flat-for-global

cdevadas replied:
The attr flat-for-global was used to force SGPR usage and eventually their spills. Otherwise, it uses GLOBAL_STORE_DWORD that only need VGPRs.

arsenm replied:
This is far too roundabout way of getting there, there must be a simpler way of getting SGPR uses. No new uses of -flat-for-global should be introduced

arsenm commented on llvm/test/CodeGen/AMDGPU/wwm-regalloc-error.ll:
```diff
@@ -0,0 +1,53 @@
+  call void asm sideeffect "", "~{v[56:63]}" ()
+  call void asm sideeffect "", "~{v[64:71]}" ()
+  call void asm sideeffect "", "~{v[72:79]}" ()
+  call void asm sideeffect "", "~{v[80:87]}" ()
+  call void asm sideeffect "", "~{v[88:95]}" ()
+  call void asm sideeffect "", "~{v[96:103]}" ()
+  call void asm sideeffect "", "~{v[104:111]}" ()
+  call void asm sideeffect "", "~{v[112:119]}" ()
+  call void asm sideeffect "", "~{v[120:127]}" ()
+  call void asm sideeffect "", "~{v[128:135]}" ()
+  call void asm sideeffect "", "~{v[136:143]}" ()
+  call void asm sideeffect "", "~{v[144:151]}" ()
+  call void asm sideeffect "", "~{v[152:159]}" ()
+  call void asm sideeffect "", "~{v[160:167]}" ()
+  call void asm sideeffect "", "~{v[168:175]}" ()
+  call void asm sideeffect "", "~{v[176:183]}" ()
+  call void asm sideeffect "", "~{v[184:191]}" ()
+  call void asm sideeffect "", "~{v[192:199]}" ()
+  call void asm sideeffect "", "~{v[200:207]}" ()
+  call void asm sideeffect "", "~{v[208:215]}" ()
+  call void asm sideeffect "", "~{v[216:223]}" ()
+  call void asm sideeffect "", "~{v[224:231]}" ()
+  call void asm sideeffect "", "~{v[232:239]}" ()
+  call void asm sideeffect "", "~{v[240:247]}" ()
+  call void asm sideeffect "", "~{v[248:255]}" ()
```
I think you should only need to go up to the number of registers that support 1024 workgroups

arsenm commented on llvm/include/llvm/CodeGen/RegAllocCommon.h:
```diff
@@ -10,22 +10,25 @@
 typedef std::function<bool(const TargetRegisterInfo &TRI,
-                           const TargetRegisterClass &RC)> RegClassFilterFunc;
+                           const MachineRegisterInfo &MRI, const Register Reg)>
+    RegClassFilterFunc;
```
Rename this to RegAllocFilterFunc?

cdevadas replied:
It is already fixed in the original PR https://github.com/llvm/llvm-project/pull/93525. I will rebase it.

arsenm commented on llvm/include/llvm/CodeGen/RegAllocCommon.h:
```diff
@@ -10,22 +10,25 @@
 
 /// Default register class filter function for register allocation. All virtual
 /// registers should be allocated.
 static inline bool allocateAllRegClasses(const TargetRegisterInfo &,
```
allocateAllRegisters

arsenm commented on llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:
```diff
@@ -1555,6 +1555,17 @@ void SIFrameLowering::determinePrologEpilogSGPRSaves(
 }
 
+// Mark all WWM VGPRs as BB LiveIns.
+static void addWwmRegBBLiveIn(MachineFunction &MF) {
```
If the registers are reserved, they don't need to be made live in. It's one or the other

cdevadas replied:
We shift back the wwm-regs to the lowest range at this point. They aren't explicitly marked reserved. I'll do that by calling MRI.reserveReg() on these new regs. With that, I can stop adding them to BBLiveIns.

arsenm commented on llvm/lib/Target/AMDGPU/SIFrameLowering.cpp:
```diff
@@ -1608,6 +1605,25 @@ void SIFrameLowering::determineCalleeSaves(MachineFunction &MF,
+    // are of 32-bit size. SIPreAllocateWWMRegs pass can add tuples into WWM
+    // reserved registers.
+    const TargetRegisterClass *RC = TRI->getPhysRegBaseClass(Reg);
+    if (TRI->getRegSizeInBits(*RC) > 32)
+      continue;
```
You can't just ignore these?

cdevadas replied:
No. As the comment indicate, `SIPreAllocateWWMRegs` adds register tuples into `WWMReservedRegs` list. They're already from the lowest range, so attempting to shift them further is unnecessary.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUReserveWWMRegs.cpp:
```diff
@@ -0,0 +1,104 @@
+  for (MachineBasicBlock &MBB : MF) {
+    for (MachineInstr &MI : MBB) {
+      for (MachineOperand &MO : MI.operands()) {
+        if (!MO.isReg())
```
Is it worth skipping operands that aren't already renamable? Is there a way we can set some property on the spill pseudos to prevent getting renamable on these operands in the first place? 

e.g. set hasExtraSrcRegAllocReq on SI_SPILL_S32_TO_VGPR

cdevadas replied:
It's not only the SGPR spill instructions, but all instances of the wwm-regs. The IMPLICIT_DEF at the dominator blocks, the wwm-reg spills and copies, etc shouldn't have the renamable flag. So hasExtraSrcRegAllocReq might not help.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUReserveWWMRegs.cpp:
```diff
@@ -0,0 +1,104 @@
+  // The renamable flag can't be set for reserved registers. Reset the flag for
+  // MOs involving wwm-regs as they will be reserved during vgpr-regalloc
+  // pipeline.
+  for (MachineBasicBlock &MBB : MF) {
```
If we know we only have 32-bit register uses, you could instead look at the use lists for the WWM reserved registers instead of walking every instruction in the function

cdevadas replied:
Will do. That would be super efficient here.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUReserveWWMRegs.cpp:
```diff
@@ -0,0 +1,104 @@
+
+        Register Reg = MO.getReg();
+        if (Reg.isPhysical() &&
+            llvm::is_contained(MFI->getWWMReservedRegs(), Reg))
```
Don't need llvm::

arsenm commented:
Title should be fixed before merge

cdevadas commented:
It depends on https://github.com/llvm/llvm-project/pull/92809. The downstream compiler would break some applications without this dependent patch.

arsenm commented:
> It depends on #92809. The downstream compiler would break some applications without this dependent patch.

It would really be better if it didn't. Did the other implementation depend on it as well?

cdevadas commented:
> > It depends on #92809. The downstream compiler would break some applications without this dependent patch.
> 
> It would really be better if it didn't. Did the other implementation depend on it as well?

The regalloc strategy changes and noted some tests (part of the extended testing) caused the assertion in SplitKit while trying to recompute the insertion point for COPY at the BB prologue. Yes, the same behavior was observed with the other implementation too.

arsenm commented:
Can this finally land after 735a5f67e351fc072a9f7c18b030036681f7935a?

cdevadas commented:
> Can this finally land after [735a5f6](https://github.com/llvm/llvm-project/commit/735a5f67e351fc072a9f7c18b030036681f7935a)?

Yes. Will be landing in a few minutes.

yxsamliu commented:
> LLVM Buildbot has detected a new failure on builder `lld-x86_64-win` running on `as-worker-93` while building `llvm` at step 7 "test-build-unified-tree-check-all".
> 
> Full details are available at: https://lab.llvm.org/buildbot/#/builders/146/builds/1277
> 
> Here is the relevant piece of the build log for the reference

The failure was due to blender perf degradation about 10%. Since this patch is critical for correctness, we won't revert this patch due to perf degradation. Instead, we will address it with following up fix. Currently this failure was silenced by rebasing the perf results in the buildbot.
